subscriptions/providers/apple_in_app/__init__.py

```diff
@@ -0,0 +1,160 @@
+from dataclasses import dataclass
+from decimal import Decimal
+from typing import (
+    ClassVar,
+    Iterable,
+    Optional,
+    Tuple,
+)
+
+from django.conf import settings
+from django.contrib.auth.base_user import AbstractBaseUser
+from django.core.exceptions import SuspiciousOperation
+from rest_framework.request import Request
+from rest_framework.response import Response
+
+from subscriptions.exceptions import InvalidOperation
+from subscriptions.models import (
+    Plan,
+    Subscription,
+    SubscriptionPayment,
+)
+from .api import (
+    AppleAppStoreAPI,
+    AppleInApp,
+    AppleVerifyReceiptResponse,
+)
+from .app_store import (
+    AppStoreNotification,
+    AppStoreNotificationTypeV2,
+    PayloadValidationError,
+    setup_original_apple_certificate,
+)
+from .. import Provider
+
+
+@dataclass
+class AppleInAppProvider(Provider):
+    transaction_receipt_tag: ClassVar[str] = 'transaction_receipt'
+    signed_payload_tag: ClassVar[str] = 'signedPayload'
+
+    codename: ClassVar[str] = 'apple_in_app'
+
+    api: AppleAppStoreAPI = None
+
+    def __post_init__(self):
+        self.api = AppleAppStoreAPI(settings.APPLE_SHARED_SECRET)
+        setup_original_apple_certificate(settings.APPLE_ROOT_CERTIFICATE_PATH)
+
+    def charge_online(self, user: AbstractBaseUser, plan: Plan, subscription: Optional[Subscription] = None,
+                      quantity: int = 1) -> Tuple[SubscriptionPayment, str]:
+        """
+        In case of in-app purchase this operation is triggered from the mobile application library.
+        """
+        raise InvalidOperation()
+
+    def charge_offline(self, user: AbstractBaseUser, plan: Plan, subscription: Optional[Subscription] = None,
+                       quantity: int = 1,
+                       reference_payment: Optional[SubscriptionPayment] = None) -> SubscriptionPayment:
+        raise InvalidOperation()
+
+    def webhook(self, request: Request, payload: dict) -> Response:
+        if self.transaction_receipt_tag in payload:
+            return self._handle_receipt(request, payload)
+        elif self.signed_payload_tag in payload:
+            return self._handle_app_store(request, payload)
+        else:
+            # Invalid, unhandled request.
+            return Response(status=400)
+
+    def check_payments(self, payments: Iterable[SubscriptionPayment]):
+        pass
```

Comment:
From my understanding we shouldn't ever get a `PENDING` payment – application should inform us about operations after they were completed and Apple will inform us about multiple events, but we currently react only to the one that means that the payment was finished.
I'll go with a warning, as this could mean that something fishy is going on.

Reply:
Done.